File: packages/core-py/src/context_shunt/accounting.py
"""Token accounting: signed, one-time-credited, and never zero-for-unknown.

Definitions, verbatim from the contract::

    main_context_tokens_saved = baseline_credit_tokens - main_model_envelope_tokens
    net_tokens_saved          = main_context_tokens_saved
                              - reader_input_tokens - reader_output_tokens

Both results are **signed**. A refined question, a failed retry and an inspect page all
produce a negative net, and that is the correct answer: they cost context without
withholding anything new.

Three rules keep the numbers honest.

**Zero is never unknown.** A token count that the provider did not report is stored as
SQL NULL and rendered as ``null``. Where a number is still required for the arithmetic,
the estimate is derived from bytes we measured ourselves and the record says
``bytes_div_4`` rather than pretending the count was exact. ``unknown`` appears only when
no attempt was made at all.

**The baseline is credited once.** ``baseline_credit_tokens`` is non-zero only on the
operation that first withheld a given snapshot. Every later operation over the same
snapshot records a zero credit and still records its own envelope and reader overhead, so
repeated recovery shows up as accumulating cost, not as repeated savings.

**A counterfactual is labelled as one.** ``full_payload_counterfactual`` is what the whole
payload *would* have cost had it entered the conversation - we measured the payload, so
the number is real, but the saving is a counterfactual. ``host_truncated_observed`` is the
different, smaller baseline that applies when the host had already truncated the result
before we saw it: crediting the full payload there would be a fabrication.

Egress is measured after the envelope is complete. The envelope carries only the opaque
``accounting_id``; the record it points at is written afterwards from the exact serialized
bytes, so the measurement can never include itself.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Any

from .limits import DEFAULT_LIMITS, Limits
from .provenance import TokenMethod
from .store import OperationRecord, new_operation_id


class OperationKind(str, Enum):
    GATE_BLOCK = "gate_block"
    CAPTURE = "capture"
    READ = "read"
    REFINED_READ = "refined_read"
    INSPECT = "inspect"
    STATS = "stats"
    SPILL = "spill"


class BaselineKind(str, Enum):
    #: We measured the entire payload; the saving is what it would have cost.
    FULL_PAYLOAD_COUNTERFACTUAL = "full_payload_counterfactual"
    #: The host truncated before we saw it; only the truncated size is observable.
    HOST_TRUNCATED_OBSERVED = "host_truncated_observed"
    #: Nothing was withheld by this operation.
    NONE = "none"


class DeliveryBoundary(str, Enum):
    ENVELOPE = "envelope"
    EXTRACTION = "extraction"
    POINTER = "pointer"
    BLOCK_MESSAGE = "block_message"
    NONE = "none"


def estimate_tokens(byte_count: int, limits: Limits = DEFAULT_LIMITS) -> int:
    """The single deterministic estimator, named ``bytes_div_4`` in every record."""
    divisor = max(1, limits.bytes_per_token_estimate)
    return (max(0, byte_count) + divisor - 1) // divisor


@dataclass(frozen=True)
class Baseline:
    """What this operation withheld from the main context."""

    kind: BaselineKind
    raw_input_bytes: int
    tokens: int | None
    method: TokenMethod

    @classmethod
    def withheld_payload(cls, raw_input_bytes: int, *, limits: Limits = DEFAULT_LIMITS) -> Baseline:
        return cls(
            kind=BaselineKind.FULL_PAYLOAD_COUNTERFACTUAL,
            raw_input_bytes=raw_input_bytes,
            tokens=estimate_tokens(raw_input_bytes, limits),
            method=TokenMethod.BYTES_DIV_4,
        )

    @classmethod
    def host_truncated(cls, observed_bytes: int, *, limits: Limits = DEFAULT_LIMITS) -> Baseline:
        """A smaller, separately labelled baseline for an already-truncated upstream."""
        return cls(
            kind=BaselineKind.HOST_TRUNCATED_OBSERVED,
            raw_input_bytes=observed_bytes,
            tokens=estimate_tokens(observed_bytes, limits),
            method=TokenMethod.BYTES_DIV_4,
        )

    @classmethod
    def none(cls) -> Baseline:
        return cls(
            kind=BaselineKind.NONE,
            raw_input_bytes=0,
            tokens=None,
            method=TokenMethod.UNKNOWN,
        )


@dataclass(frozen=True)
class ReaderCost:
    """What the reader model spent. ``None`` means the provider did not report it."""

    input_tokens: int | None = None
    output_tokens: int | None = None
    cache_tokens: int | None = None
    method: TokenMethod = TokenMethod.NOT_APPLICABLE
    attempts_started: int = 0
    attempts_usage_complete: int = 0

    @classmethod
    def none(cls) -> ReaderCost:
        return cls()

    @classmethod
    def estimated_from_bytes(
        cls,
        *,
        prompt_bytes: int,
        completion_bytes: int,
        attempts_started: int,
        limits: Limits = DEFAULT_LIMITS,
    ) -> ReaderCost:
        """Deterministic fallback when usage was not reported.

        The byte counts come from the messages this core built and the text it received,
        so the estimate is reproducible - but the record still says ``bytes_div_4`` so
        nobody mistakes it for provider truth.
        """
        return cls(
            input_tokens=estimate_tokens(prompt_bytes, limits),
            output_tokens=estimate_tokens(completion_bytes, limits),
            cache_tokens=None,
            method=TokenMethod.BYTES_DIV_4,
            attempts_started=attempts_started,
            attempts_usage_complete=0,
        )

    @property
    def charged_input(self) -> int:
        return self.input_tokens or 0

    @property
    def charged_output(self) -> int:
        return self.output_tokens or 0


@dataclass(frozen=True)
class Egress:
    """The exact bytes that crossed into the main model context."""

    boundary: DeliveryBoundary
    byte_count: int
    method: TokenMethod = TokenMethod.BYTES_DIV_4

    def tokens(self, limits: Limits = DEFAULT_LIMITS) -> int:
        return estimate_tokens(self.byte_count, limits)


def compose(
    *,
    operation_id: str,
    kind: OperationKind,
    status: str,
    code: str,
    baseline: Baseline,
    baseline_credited: bool,
    reader: ReaderCost,
    egress: Egress,
    limits: Limits = DEFAULT_LIMITS,
    credited_bytes: int | None = None,
) -> OperationRecord:
    """Build the record for one operation.

    ``baseline_credited`` is the store's one-time answer for this snapshot. When it is
    false the credit is zero even though ``raw_input_baseline_tokens`` still reports what
    the payload measures, so the record shows both "this much was withheld overall" and
    "this operation claims none of that saving".

    ``credited_bytes`` separates the two for a *mixed* selection. The measurement covers
    every selected source, but the credit may cover only some of them: a read that reuses
    an already-credited source alongside a new one withholds nothing new for the former.
    Crediting the whole measured baseline whenever any part of it was new inflated the
    reported saving on every such read. When it is omitted the credit is the whole
    baseline, which is the single-source case and the previous behaviour.
    """
    envelope_tokens = egress.tokens(limits)
    if not baseline_credited or baseline.tokens is None:
        credit = 0
    elif credited_bytes is None:
        credit = baseline.tokens
    else:
        credit = estimate_tokens(credited_bytes, limits)
    main_saved = credit - envelope_tokens
    net_saved = main_saved - reader.charged_input - reader.charged_output
    return OperationRecord(
        operation_id=operation_id,
        kind=kind.value,
        status=status,
        code=code,
        raw_input_bytes=baseline.raw_input_bytes,
        raw_input_baseline_tokens=baseline.tokens,
        baseline_kind=baseline.kind.value,
        baseline_method=_baseline_method(baseline),
        baseline_credit_tokens=credit,
        main_model_envelope_bytes=egress.byte_count,
        main_model_envelope_tokens=envelope_tokens,
        envelope_token_method=egress.method.value,
        reader_input_tokens=reader.input_tokens,
        reader_output_tokens=reader.output_tokens,
        reader_cache_tokens=reader.cache_tokens,
        reader_token_method=reader.method.value,
        attempts_started=reader.attempts_started,
        attempts_usage_complete=reader.attempts_usage_complete,
        delivery_boundary=egress.boundary.value,
        main_context_tokens_saved=main_saved,
        net_tokens_saved=net_saved,
    )


def _baseline_method(baseline: Baseline) -> str:
    """The DDL admits exact/bytes_div_4/unknown for a baseline; not_applicable is not one."""
    if baseline.method is TokenMethod.NOT_APPLICABLE:
        return TokenMethod.UNKNOWN.value
    return baseline.method.value


def totals_to_dict(raw: dict[str, Any], *, operations: int | None = None) -> dict[str, Any]:
    """Shape the store's aggregate for the stats envelope, preserving nulls."""
    out = {
        "operations": int(raw.get("operations") or 0) if operations is None else operations,
        "raw_input_bytes": int(raw.get("raw_input_bytes") or 0),
        "baseline_credit_tokens": int(raw.get("baseline_credit_tokens") or 0),
        "main_model_envelope_tokens": int(raw.get("main_model_envelope_tokens") or 0),
        "main_context_tokens_saved": int(raw.get("main_context_tokens_saved") or 0),
        "net_tokens_saved": int(raw.get("net_tokens_saved") or 0),
        "attempts_started": int(raw.get("attempts_started") or 0),
        "attempts_usage_complete": int(raw.get("attempts_usage_complete") or 0),
        "disclosed_bytes": int(raw.get("disclosed_bytes") or 0),
    }
    for key in ("reader_input_tokens", "reader_output_tokens", "reader_cache_tokens"):
        value = raw.get(key)
        # A NULL sum means no operation reported this direction. It stays null.
        out[key] = None if value is None else int(value)
    return out


__all__ = [
    "Baseline",
    "BaselineKind",
    "DeliveryBoundary",
    "Egress",
    "OperationKind",
    "ReaderCost",
    "compose",
    "estimate_tokens",
    "new_operation_id",
    "totals_to_dict",
]

File: packages/core-ts/src/accounting.ts
/**
 * Token accounting: signed, one-time-credited, and never zero-for-unknown.
 *
 * Definitions, verbatim from the contract:
 *
 *     mainContextTokensSaved = baselineCreditTokens - mainModelEnvelopeTokens
 *     netTokensSaved         = mainContextTokensSaved - readerInputTokens - readerOutputTokens
 *
 * Both results are **signed**. A refined question, a failed retry and an inspect page all
 * produce a negative net, and that is the correct answer: they cost context without
 * withholding anything new.
 *
 * Three rules keep the numbers honest.
 *
 * **Zero is never unknown.** A token count the provider did not report is stored as SQL
 * NULL and rendered as `null`. Where a number is still required for the arithmetic, the
 * estimate is derived from bytes we measured ourselves and the record says `bytes_div_4`
 * rather than pretending the count was exact. `unknown` appears only when no attempt was
 * made at all.
 *
 * **The baseline is credited once.** `baselineCreditTokens` is non-zero only on the
 * operation that first withheld a given snapshot, so repeated recovery shows up as
 * accumulating cost rather than repeated savings.
 *
 * **A counterfactual is labelled as one.** `full_payload_counterfactual` is what the whole
 * payload *would* have cost; `host_truncated_observed` is the different, smaller baseline
 * that applies when the host had already truncated before we saw it. Crediting the full
 * payload there would be a fabrication.
 *
 * Egress is measured after the envelope is complete. The envelope carries only the opaque
 * `accounting_id`; the record it points at is written afterwards from the exact serialized
 * bytes, so the measurement can never include itself.
 */
import { randomBytes } from "node:crypto";

import { DEFAULT_LIMITS, Limits } from "./limits.js";
import type { TokenMethod } from "./provenance.js";
import type { OperationRecord } from "./store.js";

export type OperationKind =
  | "gate_block"
  | "capture"
  | "read"
  | "refined_read"
  | "inspect"
  | "stats"
  | "spill";

export type BaselineKind =
  /** We measured the entire payload; the saving is what it would have cost. */
  | "full_payload_counterfactual"
  /** The host truncated before we saw it; only the truncated size is observable. */
  | "host_truncated_observed"
  /** Nothing was withheld by this operation. */
  | "none";

export type DeliveryBoundary = "envelope" | "extraction" | "pointer" | "block_message" | "none";

export function newOperationId(): string {
  return "acc_" + randomBytes(8).toString("hex");
}

/** The single deterministic estimator, named `bytes_div_4` in every record. */
export function estimateTokens(byteCount: number, limits: Limits = DEFAULT_LIMITS): number {
  const divisor = Math.max(1, limits.bytesPerTokenEstimate);
  return Math.ceil(Math.max(0, byteCount) / divisor);
}

export interface Baseline {
  readonly kind: BaselineKind;
  readonly rawInputBytes: number;
  readonly tokens: number | undefined;
  readonly method: TokenMethod;
}

export function withheldPayloadBaseline(
  rawInputBytes: number,
  limits: Limits = DEFAULT_LIMITS,
): Baseline {
  return {
    kind: "full_payload_counterfactual",
    rawInputBytes,
    tokens: estimateTokens(rawInputBytes, limits),
    method: "bytes_div_4",
  };
}

/** A smaller, separately labelled baseline for an already-truncated upstream. */
export function hostTruncatedBaseline(
  observedBytes: number,
  limits: Limits = DEFAULT_LIMITS,
): Baseline {
  return {
    kind: "host_truncated_observed",
    rawInputBytes: observedBytes,
    tokens: estimateTokens(observedBytes, limits),
    method: "bytes_div_4",
  };
}

export function noBaseline(): Baseline {
  return { kind: "none", rawInputBytes: 0, tokens: undefined, method: "unknown" };
}

export interface ReaderCost {
  readonly inputTokens: number | undefined;
  readonly outputTokens: number | undefined;
  readonly cacheTokens: number | undefined;
  readonly method: TokenMethod;
  readonly attemptsStarted: number;
  readonly attemptsUsageComplete: number;
}

export function noReaderCost(): ReaderCost {
  return {
    inputTokens: undefined,
    outputTokens: undefined,
    cacheTokens: undefined,
    method: "not_applicable",
    attemptsStarted: 0,
    attemptsUsageComplete: 0,
  };
}

/**
 * Deterministic fallback when usage was not reported. The byte counts come from the
 * messages this core built and the text it received, so the estimate is reproducible - but
 * the record still says `bytes_div_4` so nobody mistakes it for provider truth.
 */
export function estimatedReaderCost(opts: {
  promptBytes: number;
  completionBytes: number;
  attemptsStarted: number;
  limits?: Limits;
}): ReaderCost {
  const limits = opts.limits ?? DEFAULT_LIMITS;
  return {
    inputTokens: estimateTokens(opts.promptBytes, limits),
    outputTokens: estimateTokens(opts.completionBytes, limits),
    cacheTokens: undefined,
    method: "bytes_div_4",
    attemptsStarted: opts.attemptsStarted,
    attemptsUsageComplete: 0,
  };
}

export interface Egress {
  readonly boundary: DeliveryBoundary;
  readonly byteCount: number;
  readonly method: TokenMethod;
}

export function envelopeEgress(boundary: DeliveryBoundary, byteCount: number): Egress {
  return { boundary, byteCount, method: "bytes_div_4" };
}

/** The DDL admits exact/bytes_div_4/unknown for a baseline; not_applicable is not one. */
function baselineMethod(baseline: Baseline): TokenMethod {
  return baseline.method === "not_applicable" ? "unknown" : baseline.method;
}

export interface ComposeInput {
  operationId: string;
  kind: OperationKind;
  status: string;
  code: string;
  baseline: Baseline;
  /**
   * The store's one-time answer for this snapshot. When false the credit is zero even
   * though `raw_input_baseline_tokens` still reports what the payload measures, so the
   * record shows both "this much was withheld overall" and "this operation claims none of
   * that saving".
   */
  baselineCredited: boolean;
  /**
   * How many withheld bytes this operation may claim, when that is narrower than the
   * measured baseline. The measurement covers every selected source; the credit may cover
   * only some of them, because a read that reuses an already-credited source withholds
   * nothing new for it. Crediting the whole measured baseline whenever any part was new
   * inflated the saving on every mixed-source read. Omitted means "all of it", which is
   * the single-source case and the previous behaviour.
   */
  creditedBytes?: number | undefined;
  reader: ReaderCost;
  egress: Egress;
  limits?: Limits;
}

export function composeRecord(input: ComposeInput): OperationRecord {
  const limits = input.limits ?? DEFAULT_LIMITS;
  const envelopeTokens = estimateTokens(input.egress.byteCount, limits);
  const credit =
    !input.baselineCredited || input.baseline.tokens === undefined
      ? 0
      : input.creditedBytes === undefined
        ? input.baseline.tokens
        : estimateTokens(input.creditedBytes, input.limits);
  const mainSaved = credit - envelopeTokens;
  const netSaved = mainSaved - (input.reader.inputTokens ?? 0) - (input.reader.outputTokens ?? 0);
  return {
    operationId: input.operationId,
    kind: input.kind,
    status: input.status,
    code: input.code,
    rawInputBytes: input.baseline.rawInputBytes,
    rawInputBaselineTokens: input.baseline.tokens,
    baselineKind: input.baseline.kind,
    baselineMethod: baselineMethod(input.baseline),
    baselineCreditTokens: credit,
    mainModelEnvelopeBytes: input.egress.byteCount,
    mainModelEnvelopeTokens: envelopeTokens,
    envelopeTokenMethod: input.egress.method,
    readerInputTokens: input.reader.inputTokens,
    readerOutputTokens: input.reader.outputTokens,
    readerCacheTokens: input.reader.cacheTokens,
    readerTokenMethod: input.reader.method,
    attemptsStarted: input.reader.attemptsStarted,
    attemptsUsageComplete: input.reader.attemptsUsageComplete,
    deliveryBoundary: input.egress.boundary,
    mainContextTokensSaved: mainSaved,
    netTokensSaved: netSaved,
  };
}

export interface StatsTotalsShape {
  operations: number;
  raw_input_bytes: number;
  baseline_credit_tokens: number;
  main_model_envelope_tokens: number;
  reader_input_tokens: number | null;
  reader_output_tokens: number | null;
  reader_cache_tokens: number | null;
  main_context_tokens_saved: number;
  net_tokens_saved: number;
  attempts_started: number;
  attempts_usage_complete: number;
  disclosed_bytes: number;
}

export interface RawTotals {
  operations: number;
  rawInputBytes: number;
  baselineCreditTokens: number;
  mainModelEnvelopeTokens: number;
  readerInputTokens: number | undefined;
  readerOutputTokens: number | undefined;
  readerCacheTokens: number | undefined;
  mainContextTokensSaved: number;
  netTokensSaved: number;
  attemptsStarted: number;
  attemptsUsageComplete: number;
  disclosedBytes: number;
}

/** Shape the store's aggregate for the stats envelope, preserving nulls. */
export function totalsToShape(raw: RawTotals): StatsTotalsShape {
  return {
    operations: raw.operations,
    raw_input_bytes: raw.rawInputBytes,
    baseline_credit_tokens: raw.baselineCreditTokens,
    main_model_envelope_tokens: raw.mainModelEnvelopeTokens,
    // A null sum means no operation reported this direction. It stays null.
    reader_input_tokens: raw.readerInputTokens ?? null,
    reader_output_tokens: raw.readerOutputTokens ?? null,
    reader_cache_tokens: raw.readerCacheTokens ?? null,
    main_context_tokens_saved: raw.mainContextTokensSaved,
    net_tokens_saved: raw.netTokensSaved,
    attempts_started: raw.attemptsStarted,
    attempts_usage_complete: raw.attemptsUsageComplete,
    disclosed_bytes: raw.disclosedBytes,
  };
}

export interface OperationRecordShape {
  operation_id: string;
  kind: string;
  status: string;
  code: string;
  raw_input_bytes: number;
  raw_input_baseline_tokens: number | null;
  baseline_kind: string;
  baseline_method: string;
  baseline_credit_tokens: number;
  main_model_envelope_bytes: number;
  main_model_envelope_tokens: number;
  envelope_token_method: string;
  reader_input_tokens: number | null;
  reader_output_tokens: number | null;
  reader_cache_tokens: number | null;
  reader_token_method: string;
  attempts_started: number;
  attempts_usage_complete: number;
  delivery_boundary: string;
  main_context_tokens_saved: number;
  net_tokens_saved: number;
}

export function recordToShape(record: OperationRecord): OperationRecordShape {
  return {
    operation_id: record.operationId,
    kind: record.kind,
    status: record.status,
    code: record.code,
    raw_input_bytes: record.rawInputBytes,
    raw_input_baseline_tokens: record.rawInputBaselineTokens ?? null,
    baseline_kind: record.baselineKind,
    baseline_method: record.baselineMethod,
    baseline_credit_tokens: record.baselineCreditTokens,
    main_model_envelope_bytes: record.mainModelEnvelopeBytes,
    main_model_envelope_tokens: record.mainModelEnvelopeTokens,
    envelope_token_method: record.envelopeTokenMethod,
    reader_input_tokens: record.readerInputTokens ?? null,
    reader_output_tokens: record.readerOutputTokens ?? null,
    reader_cache_tokens: record.readerCacheTokens ?? null,
    reader_token_method: record.readerTokenMethod,
    attempts_started: record.attemptsStarted,
    attempts_usage_complete: record.attemptsUsageComplete,
    delivery_boundary: record.deliveryBoundary,
    main_context_tokens_saved: record.mainContextTokensSaved,
    net_tokens_saved: record.netTokensSaved,
  };
}
